File: python/python-sdk/kestrapy/api/tenant_access_api.py
from typing import List, Optional

from kestrapy.base_api import BaseApi
from kestrapy.models.iam_tenant_access_controller_api_create_tenant_access_request import (
    IAMTenantAccessControllerApiCreateTenantAccessRequest,
)
from kestrapy.models.iam_tenant_access_controller_api_tenant_access import (
    IAMTenantAccessControllerApiTenantAccess,
)
from kestrapy.models.iam_tenant_access_controller_api_user_tenant_access import (
    IAMTenantAccessControllerApiUserTenantAccess,
)
from kestrapy.models.paged_results_iam_tenant_access_controller_api_user_tenant_access import (
    PagedResultsIAMTenantAccessControllerApiUserTenantAccess,
)
from kestrapy.models.query_filter import QueryFilter


class TenantAccessApi(BaseApi):
    """Tenant access (`/api/v1/{tenant}/tenant-access`).

    Since Kestra 2.0 tenant access is a prerequisite rather than a side effect:
    a user must be granted access before it can be added to a group.
    """

    def list_tenant_access(
        self,
        tenant: str,
        page: Optional[int] = None,
        size: Optional[int] = None,
        sort: Optional[List[str]] = None,
        filters: Optional[List[QueryFilter]] = None,
    ) -> PagedResultsIAMTenantAccessControllerApiUserTenantAccess:
        """The users that can reach the tenant. GET /api/v1/{tenant}/tenant-access."""
        path = self._tenant_path(tenant, "tenant-access")
        params = list(self._build_query_params(page=page, size=size).items())
        self._append_repeated_param(params, "sort", sort)
        self._append_filter_params(params, filters)
        return self._json_request(
            "GET", path, PagedResultsIAMTenantAccessControllerApiUserTenantAccess, params=params
        )

    def tenant_access(
        self, user_id: str, tenant: str
    ) -> IAMTenantAccessControllerApiTenantAccess:
        """One user's access to the tenant.
        GET /api/v1/{tenant}/tenant-access/{userId}."""
        path = self._tenant_path(tenant, "tenant-access", user_id)
        return self._json_request(
            "GET", path, IAMTenantAccessControllerApiTenantAccess
        )

    def create_tenant_access(self, user_id: str, tenant: str) -> None:
        """Grant access by user id. The server answers 201 with no body.
        PUT /api/v1/{tenant}/tenant-access/{userId}."""
        path = self._tenant_path(tenant, "tenant-access", user_id)
        self._void_request("PUT", path)

    def create_tenant_access_by_email(
        self, tenant: str, request: IAMTenantAccessControllerApiCreateTenantAccessRequest
    ) -> None:
        """Grant access by email. The server answers 204 with no body, and 409
        when the user already has access. POST /api/v1/{tenant}/tenant-access."""
        path = self._tenant_path(tenant, "tenant-access")
        self._void_request("POST", path, body=request)

    def delete_tenant_access(self, user_id: str, tenant: str) -> None:
        """Revoke a user's access to the tenant.
        DELETE /api/v1/{tenant}/tenant-access/{userId}."""
        path = self._tenant_path(tenant, "tenant-access", user_id)
        self._void_request("DELETE", path)

    def autocomplete_tenant_access(
        self, tenant: str, request: dict
    ) -> List[IAMTenantAccessControllerApiUserTenantAccess]:
        """Resolve users reachable from the tenant.
        POST /api/v1/{tenant}/tenant-access/autocomplete."""
        path = self._tenant_path(tenant, "tenant-access", "autocomplete")
        return self._json_list_request(
            "POST", path, IAMTenantAccessControllerApiUserTenantAccess, body=request
        )

File: go-sdk/kestra_api_client/tenant_access_api.go
package kestra_api_client

import "context"

// TenantAccessAPI covers /api/v1/{tenant}/tenant-access.
//
// Since Kestra 2.0 tenant access is a prerequisite rather than something granted
// as a side effect: adding a user to a group resolves the user through a
// hasTenantAccess filter and rejects it otherwise, so the access has to be
// granted first through one of the Create calls here.
type TenantAccessAPI struct {
	baseAPI
}

// ListTenantAccess returns the users that can reach the tenant.
func (a *TenantAccessAPI) ListTenantAccess(ctx context.Context, tenant string, page, size *int, sort []string, filters []SearchFilter) (*PagedResultsIAMTenantAccessControllerApiUserTenantAccess, error) {
	params := buildQueryParams("page", page, "size", size)
	appendRepeatedParam(params, "sort", sort)
	appendFilterParams(params, filters)
	return doJSON[*PagedResultsIAMTenantAccessControllerApiUserTenantAccess](&a.baseAPI, ctx, "GET", tenantPath(tenant, "tenant-access"), nil, params)
}

// TenantAccess returns one user's access to the tenant.
func (a *TenantAccessAPI) TenantAccess(ctx context.Context, userId, tenant string) (*IAMTenantAccessControllerApiTenantAccess, error) {
	return doJSON[*IAMTenantAccessControllerApiTenantAccess](&a.baseAPI, ctx, "GET", tenantPath(tenant, "tenant-access", userId), nil, nil)
}

// CreateTenantAccess grants access by user id. The server answers 201 with no body.
func (a *TenantAccessAPI) CreateTenantAccess(ctx context.Context, userId, tenant string) error {
	return a.doVoidJSON(ctx, "PUT", tenantPath(tenant, "tenant-access", userId), nil, nil)
}

// CreateTenantAccessByEmail grants access by email. The server answers 204 with no
// body, and 409 when the user already has access to the tenant.
func (a *TenantAccessAPI) CreateTenantAccessByEmail(ctx context.Context, tenant string, request interface{}) error {
	return a.doVoidJSON(ctx, "POST", tenantPath(tenant, "tenant-access"), request, nil)
}

// DeleteTenantAccess revokes a user's access to the tenant.
func (a *TenantAccessAPI) DeleteTenantAccess(ctx context.Context, userId, tenant string) error {
	return a.doVoidJSON(ctx, "DELETE", tenantPath(tenant, "tenant-access", userId), nil, nil)
}

// AutocompleteTenantAccess resolves users reachable from the tenant.
func (a *TenantAccessAPI) AutocompleteTenantAccess(ctx context.Context, tenant string, request interface{}) ([]IAMTenantAccessControllerApiUserTenantAccess, error) {
	return doJSON[[]IAMTenantAccessControllerApiUserTenantAccess](&a.baseAPI, ctx, "POST", tenantPath(tenant, "tenant-access", "autocomplete"), request, nil)
}
